### DISCUSSION TITLE
Questions About the Multivariate Random Forest in the Package

### DISCUSSION THREAD

**Discussion by novawz** [kogalur/randomForestSRC#417] [Q&A]
Hello,

I hope this message finds you well. I'm currently developing an R package using the randomForestSRC as a foundation. I've encountered a couple of questions related to the multivariate random forests in the function. I was hoping you might be able to provide some guidance:

1. For response variables with high dimensionality, **is there a parameter in the `rfsrc()` function that selects a small subset of variables in the response for each tree, either randomly or based on weights?** I noticed the `ytry` and `yvar.wt` parameters. However, it seems that `ytry` only works with unsupervised random forests to select ytry numbers of variables in predictors as the outcome. I'm unclear about how to use the `yvar.wt` functions. I tried assigning weights to variables in the response using this parameter, but it appears nothing changed. 

2. I'm also considering the possibility of assigning weights to each variable in the outcome in the splitting statistics. As the multivariate splitting rule sums all response variables ($D = \sum D_j$, where $D_j$ is the splitting statistic for variable $j$ in the outcome), **would it be possible to assign weights ($w_j$) to the $D_j$ ($D = \sum w_jD_j$) so that I can prioritize more important variables in the outcome when calculating splitting statistics?** I think the parameter `yvar.wt` may be capable of doing this, but I'm not quite sure how to use it.

Any insights you could provide would be greatly appreciated. Thank you!

**Comment by ishwaran** [comment]
The `ytry` option only applies to unsupervised forests and the `yvar.wt` option is used for developing custom splitting rules.  Unfortunately, the package currently does not have a way to randomly select subsets of the responses for node splitting in multivariate regression.  It's something we have thought about before and it's on our list of things to do.

Regarding your second question, this sounds like you want to develop a custom splitting rule (which would involve using `yvar.wt`).  Take a look at this [discussion](https://github.com/kogalur/randomForestSRC/discussions/197) and see if that helps.

**Reply by novawz** [reply]
Thank you for your response, Dr. Ishwaran. After reading your reply, I have a few additional questions:

1. To confirm my understanding of this process, could we assign `splitrule = "custom1"` and use `yvar.wt` to set the weights, thereby assigning the split weight $w_i$ to the outcome variable $Y_i$?
2. In the [discussion](https://github.com/kogalur/randomForestSRC/discussions/197) for the `yvar.wt` parameter, an example is given where `yvar.wt = c(1,0)` is used to input only the first outcome variable into the split. Could we instead assign, for example, `yvar.wt = c(0.2,0.8)` to multiply the first variable's splitting statistics by 0.2 and the second variable's splitting statistics by 0.8?
3. After setting `yvar.wt` and assigning `splitrule` to `"custom1"` to fit the model, does the model object contain a return value reflecting the `yvar.wt` assignment?
4. Regarding the first question, is it possible to select a subset of outcomes in each tree or a limited number of trees in the forest? I tried to implement this, but it was quite slow since parallel execution should not be used around the `rfsrc()` function, as noted in the Vignette.

Thank you again for your time and help!

**Comment by kogalur** [comment]
I'd like to return to the original query.  I went back and looked at the code. We may have spoken too soon.  It looks like we put in support for ytry and yvar.wt for the multivariate case in release 3.1.1 (July, 2022).  You should convince yourself that the following example is working:

```
## mtcars based multivariate example
data(mtcars)

## add fake responses
mult.mtcars  <- cbind(mtcars, mtcars$mpg, mtcars$mpg)
names(mult.mtcars) = c(names(mtcars), "mpg2", "mpg3")

## noise up the fake responses
mult.mtcars$mpg2 <- sample(mtcars$mpg)
mult.mtcars$mpg3 <- sample(mtcars$mpg)

formula = as.formula(Multivar(mpg, mpg2, mpg3)  ~ .)

## select 2 of the 3 responses randomly at each split with an associated weight vector.
yvar.wt = c(1.0, 0.5, 0.25)
ytry = 2

mult.grow <- rfsrc(formula = formula, data = mult.mtcars, ytry = ytry, yvar.wt = yvar.wt)

## Also, compare the following two results, as they should be similar:
yvar.wt = c(1.0, 00000.1, 00000.1)
ytry = 1

result1 = rfsrc(formula = formula, data = mult.mtcars, ytry = ytry, yvar.wt = yvar.wt)
result2 = rfsrc(mpg ~ ., mtcars)

```

This should address Point (1) of your question.  

As for Point (2), it is quite separate and distinct from Point (1).  Clearly, `yvar.wt` prioritizes a response variable, only when `ytry < r` (the number of responses). If you set `ytry << r` your split statistic will favour the heavy weighted responses.  But if `r` is large and `ytry ~= r`, you will use almost all responses in the split statistic, and it will be the equivalent of not using a weight vector at all.  

If you want to modify the split statistic directly according to `yvar.wt` in the same way that we modify the split statistic using `split.wt`, you cannot do that with the current package.  Recall that the vector `split.wt` multiplies the natural split statistic directly by a weight associated with the corresponding x-variable.  To accomplish Point (2) you'd have to modify the source package.  I could _probably_ give you a line number and a few lines of code that would accomplish this hack if you really wanted to try it, but there are no guarantees.

**Reply by novawz** [reply]
Thank you for your detailed explanation, Kogalur. If I understand correctly from your example, would it be accurate to say that `yvar.wt` represents the predefined probability of selecting the outcome variables for splitting when `ytry` is set to a value less than `r`? 
Since the default value of `ytry` is set to NULL, if we assign values to `yvar.wt` such as `yvar.wt = c(1.0, 0.5, 0.25)`, and leave `ytry` at its default value of NULL, would this still be equivalent to the original multivariate splitting rules on all three variables?
From the previous [discussion](https://github.com/kogalur/randomForestSRC/discussions/197), you mentioned that we need to set the `splitrule` to "custom" to assign the `yvar.wt`. Is this still the case, or can we leave the `splitrule` as the default?
Thank you again for your time and help!

**Comment by kogalur** [comment]
I believe you can do what you want using the default multivariate split rule.  I wouldn't concern myself with the custom split rule at this point in time.  Note that `yvar.wt` and `ytry` work together in the split rule. `ytry` represents sampling without replacement.  `yvar.wt` defines the probability density function.  On each selection of a y-variable , we modify the PDF and remove the slot corresponding to the the y-variable selected.  Clearly, if the distribution is uniform, this just reduces to the R-function `sample(r, ytry)`. I wouldn't leave any parameter at their default value if you are triggering the pair (yvar.wt, ytry)`. Tell the R-side exactly what you want.

**Reply by novawz** [reply]
Thank you for the clarification, Kogalur. This is extremely helpful!
